src/instrumentationESM.ts

```diff
@@ -0,0 +1,148 @@
+        );
+
+        return instrumentationNodeModuleDefinition;
+    }
+
+    /**
+     * Internal method to patch Azure Functions SDK.
+     * Sets up logging and trace context propagation hooks.
+     * @param azFunc Azure Functions module exports.
+     * @returns Patched Azure Functions module exports.
+     */
+    private _patch(azFunc: typeof AzFunc): typeof AzFunc {
+        // Inform Azure Functions Host that logs will be sent directly from Node.js
+        azFunc.app.setup({
+            capabilities: {
+                WorkerOpenTelemetryEnabled: true,
+            },
+        });
+
+        // Hook into Azure Functions logging to emit logs via OpenTelemetry
+        const logDisposable = azFunc.app.hook.log((context) => {
+            this.logger.emit({
+                body: context.message,
+                severityNumber: toOtelSeverityNumber(context.level),
+                severityText: context.level,
+            });
+        });
+
+        // Hook into Azure Functions pre-invocation to propagate trace context
+        const preInvocationDisposable = azFunc.app.hook.preInvocation((context) => {
+            const traceContext = context.invocationContext.traceContext;
+            if (traceContext) {
+                // Extract trace context from Azure Functions invocation context
+                const extractedContext = propagation.extract(otelContext.active(), {
+                    traceparent: traceContext.traceParent,
+                    tracestate: traceContext.traceState,
+                });
+
+                // Bind the extracted context to the function handler
+                context.functionHandler = otelContext.bind(extractedContext, context.functionHandler);
+            }
+        });
+
+        // Store disposables for later cleanup
+        this.azFuncDisposable.push(logDisposable, preInvocationDisposable);
+
+        return azFunc;
+    }
+
+    /**
+     * Internal method to unpatch Azure Functions SDK.
+     * Cleans up hooks and resets capabilities.
+     * @param azFunc Azure Functions module exports.
+     */
+    private _unPatch(azFunc: typeof AzFunc): void {
+        // Dispose of all registered hooks
+        if (this.azFuncDisposable) {
+            this.azFuncDisposable.forEach((disposable: AzFunc.Disposable): void => {
+                disposable.dispose();
+            });
+            this.azFuncDisposable = [];
+        }
+
+        // Reset Azure Functions Host capabilities
+        azFunc.app.setup({
+            capabilities: {
+                WorkerOpenTelemetryEnabled: false,
+            },
+        });
+    }
+
+    /**
+     * Public method to manually trigger patching for ESM scenarios.
+     * Useful when automatic instrumentation doesn't trigger due to ESM module loading.
+     * @param azFunc Azure Functions module exports.
+     */
+    registerAzFunc(azFunc: typeof AzFunc): void {
+        this._patch(azFunc);
```

Comment:
This could lead to duplicate hooks right if the auto instrumentation triggers. That can cause duplicate logs.
We might want to check if its already patched before or unpatch before patching again

Reply:
No, this class is intended to be only when customer has the ESM app. The issue is auto instrumentation hooks (hooks does not work for ESM due to https://github.com/open-telemetry/opentelemetry-js/blob/main/doc/esm-support.md ). Hence this is the work around for customers to unblock themself until Opentelemetry address the issue.

Reply:
What will happen if otel fixes their bug and auto instrumentation starts working?

Reply:
That is correct oservation.

We are asking customer to pin the version of instrumentation in their library to avoid this (In our sample we will mention that explicitly). Also working with Hector(Opentelemtry) to pursue them for a major release